// @appium/test-support/build/lib/mock-utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MockStore = void 0;
exports.verifyMocks = verifyMocks;
exports.withMocks = withMocks;

require("source-map-support/register");

var _bluebird = _interopRequireDefault(require("bluebird"));

var _sinon = _interopRequireDefault(require("sinon"));

function withMocks(mockDefs, fn) {
  return () => {
    const mocks = new MockStore();
    beforeEach(function withMocksBeforeEach() {
      mocks.createMocks(mockDefs);
    });
    afterEach(function withMocksAfterEach() {
      mocks.reset();
    });
    fn(mocks);
  };
}

function verifyMocks(mocks) {
  mocks.verify();
}

class MockStore {
  sandbox;
  #mocks;

  constructor(sandbox) {
    this.sandbox = sandbox;
  }

  createMocks(mockDefs) {
    if (this.#mocks) {
      throw new ReferenceError('Cannot create mocks twice; call `reset()` first.');
    }

    this.sandbox = this.sandbox ?? _sinon.default.createSandbox().usingPromise(_bluebird.default);

    for (const [key, value] of Object.entries(mockDefs)) {
      this[key] = this.sandbox.mock(value);
    }

    this.#mocks = mockDefs;
    return this;
  }

  verify() {
    if (!this.sandbox) {
      throw new ReferenceError('Cannot verify mocks before they are created; call `createMocks()` first');
    }

    this.sandbox.verify();
  }

  reset() {
    var _this$sandbox;

    for (const key of Object.keys(this.#mocks ?? {})) {
      delete this[key];
    }

    (_this$sandbox = this.sandbox) === null || _this$sandbox === void 0 ? void 0 : _this$sandbox.restore();
    this.#mocks = undefined;
  }

}

exports.MockStore = MockStore;
