// @appium/test-support/build/lib/sandbox-utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SandboxStore = void 0;
exports.verifySandbox = verifySandbox;
exports.withSandbox = withSandbox;

require("source-map-support/register");

var _sinon = _interopRequireDefault(require("sinon"));

var _lodash = _interopRequireDefault(require("lodash"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _mockUtils = require("./mock-utils");

function withSandbox(mockDefs, fn) {
  if (!_lodash.default.isEmpty(mockDefs.mocks)) {
    mockDefs = mockDefs.mocks;
  }

  return () => {
    const sbx = new SandboxStore();
    beforeEach(function beforeEach() {
      sbx.createSandbox(mockDefs);
    });
    afterEach(function afterEach() {
      sbx.reset();
    });
    fn(sbx);
  };
}

function verifySandbox(sbxOrMocks) {
  sbxOrMocks.verify();
}

class SandboxStore {
  mocks;
  sandbox;

  constructor(sandbox) {
    this.sandbox = sandbox;
  }

  createSandbox(mocks = {}) {
    this.sandbox = this.sandbox ?? _sinon.default.createSandbox().usingPromise(_bluebird.default);
    this.mocks = new _mockUtils.MockStore(this.sandbox).createMocks(mocks);
  }

  verify() {
    if (!this.sandbox) {
      throw new ReferenceError('Cannot verify mocks before they are created; call `createMocks()` first');
    }

    this.sandbox.verify();
  }

  reset() {
    var _this$mocks;

    (_this$mocks = this.mocks) === null || _this$mocks === void 0 ? void 0 : _this$mocks.reset();
    delete this.sandbox;
  }

}

exports.SandboxStore = SandboxStore;
