// @appium/test-support/build/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  stubEnv: true,
  stubLog: true,
  fakeTime: true,
  withMocks: true,
  verifyMocks: true,
  withSandbox: true,
  verifySandbox: true,
  pluginE2EHarness: true,
  driverUnitTestSuite: true,
  driverE2ETestSuite: true,
  createSessionHelpers: true
};
Object.defineProperty(exports, "createSessionHelpers", {
  enumerable: true,
  get: function () {
    return _driverE2eSuite.createSessionHelpers;
  }
});
Object.defineProperty(exports, "driverE2ETestSuite", {
  enumerable: true,
  get: function () {
    return _driverE2eSuite.driverE2ETestSuite;
  }
});
Object.defineProperty(exports, "driverUnitTestSuite", {
  enumerable: true,
  get: function () {
    return _driverUnitSuite.driverUnitTestSuite;
  }
});
Object.defineProperty(exports, "fakeTime", {
  enumerable: true,
  get: function () {
    return _timeUtils.fakeTime;
  }
});
Object.defineProperty(exports, "pluginE2EHarness", {
  enumerable: true,
  get: function () {
    return _pluginE2eHarness.pluginE2EHarness;
  }
});
Object.defineProperty(exports, "stubEnv", {
  enumerable: true,
  get: function () {
    return _envUtils.stubEnv;
  }
});
Object.defineProperty(exports, "stubLog", {
  enumerable: true,
  get: function () {
    return _logUtils.stubLog;
  }
});
Object.defineProperty(exports, "verifyMocks", {
  enumerable: true,
  get: function () {
    return _mockUtils.verifyMocks;
  }
});
Object.defineProperty(exports, "verifySandbox", {
  enumerable: true,
  get: function () {
    return _sandboxUtils.verifySandbox;
  }
});
Object.defineProperty(exports, "withMocks", {
  enumerable: true,
  get: function () {
    return _mockUtils.withMocks;
  }
});
Object.defineProperty(exports, "withSandbox", {
  enumerable: true,
  get: function () {
    return _sandboxUtils.withSandbox;
  }
});

require("source-map-support/register");

require("./unhandled-rejection");

var _envUtils = require("./env-utils");

var _logUtils = require("./log-utils");

var _timeUtils = require("./time-utils");

var _mockUtils = require("./mock-utils");

var _sandboxUtils = require("./sandbox-utils");

var _pluginE2eHarness = require("./plugin-e2e-harness");

var _driverUnitSuite = require("./driver-unit-suite");

var _driverE2eSuite = require("./driver-e2e-suite");

var _helpers = require("./helpers");

Object.keys(_helpers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _helpers[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _helpers[key];
    }
  });
});
